torchvision/utils.py

```diff
@@ -300,6 +301,83 @@ def draw_segmentation_masks(
     return out.to(out_dtype)
 
 
+@torch.no_grad()
+def draw_keypoints(
+    image: torch.Tensor,
+    keypoints: torch.Tensor,
+    connectivity: Optional[Tuple[Tuple[int, int]]] = None,
+    colors: Optional[Union[List[Union[str, Tuple[int, int, int]]], str, Tuple[int, int, int]]] = None,
+    radius: int = 2,
+    width: int = 3,
+) -> torch.Tensor:
+
+    """
+    Draws Keypoints on given RGB image.
+    The values of the input image should be uint8 between 0 and 255.
+
+    Args:
+        image (Tensor): Tensor of shape (3, H, W) and dtype uint8.
+        keypoints (Tensor): Tensor of shape (num_instances, K, 2) the K keypoints location for each of the N instances,
+            in the format [x, y].
+        connectivity (Tuple[Tuple[int, int]]]): A Tuple of tuple where,
+            each tuple contains pair of keypoints to be connected.
+        colors (Union[List[Union[str, Tuple[int, int, int]]], str, Tuple[int, int, int]]): List containing the colors
+            or a single color for all the keypoints.
+            The colors can be represented as `str` or `Tuple[int, int, int]`.
+        radius (int): Integer denoting radius of keypoint.
+        width (int): Integer denoting width of line connecting keypoints.
+
+    Returns:
+        img (Tensor[C, H, W]): Image Tensor of dtype uint8 with keypoints drawn.
+    """
+
+    if not isinstance(image, torch.Tensor):
+        raise TypeError(f"The image must be a tensor, got {type(image)}")
+    elif image.dtype != torch.uint8:
+        raise ValueError(f"The image dtype must be uint8, got {image.dtype}")
+    elif image.dim() != 3:
+        raise ValueError("Pass individual images, not batches")
+    elif image.size()[0] != 3:
+        raise ValueError("Pass an RGB image. Other Image formats are not supported")
+
+    if keypoints.ndim != 3:
+        raise ValueError("keypoints must be of shape (num_instances, K, 2)")
+
+    ndarr = image.permute(1, 2, 0).numpy()
+    img_to_draw = Image.fromarray(ndarr)
+    draw = ImageDraw.Draw(img_to_draw)
+    out_dtype = torch.uint8
+
+    img_kpts = keypoints.to(torch.int64).tolist()
+
+    for kpt_inst in img_kpts:
+        for kpt in kpt_inst:
+            x1 = kpt[0] - radius
+            x2 = kpt[0] + radius
+            y1 = kpt[1] - radius
+            y2 = kpt[1] + radius
+
+            if isinstance(colors, str) or isinstance(colors, tuple):
+                draw.ellipse([x1, y1, x2, y2], fill=colors, outline=None, width=0)
```

Comment:
If we decide to pass tuple containing colors from each keypoint. It would be simply the ith index of tuple.

If we decide for colors per instance. It would need refactor.

Reply:
Yes, I propose that we handle this with a slight change.
We convert the colors list to a `Tensor`, perform the proper reshape + broadcast so that `colors` is always `(num_instances, K, 3)` tensor, and then convert it back to a list (if needed).

Reply:
I need little bit more clarity in colors. I'm assuming we are not going to specify color for every keypoint id.

Say we have 2 instances and 3 keypoints.

So what are choices possible to user?

```
# Specify one single color for all keypoints of all instances
colors = "red"

# Specify one single color for all keypoints of each instance.
colors = ["red", blue"]

# Specify color for every keypoint of every instance.

colors = [["red", "blue", "green"], ["red", "blue", "green"]]

```
Note that string can be replaced by RGB Tuple too.

Am I aligning with your thoughts?

Reply:
Hi @oke-aditya 

This logic is what I proposed. I think using broadcasting is easier to implement, and handle extra cases like if the user wants all the keypoints to have the same color, but they being different colors for each instance.

They would only need to do the following:
```python
# there are N keypoints, and 3 instances
colors = [["red"], ["green"], ["blue"]]
```

The implementation I provided above should already handle those cases without problems, let me know if you disagree.

Reply:
Somehow I'm on the edge here. :sweat_smile: The above case is handled by current implementation.
user could do. 

```
# there are N keypoints, and 3 instances.
# The first instance gets red, second gets green and third blue

colors = ["red", "blue", "green"]
```

Probably I should have attached one run with this.
To implement with numpy broadcasting vs this one. I'm not really sure which is easier (I felt current one is easier)

I had few doubts on the numpy implementation

`colors=[["red"], ["green"], ["blue"]]`

Does this mean there are 3 instances with multiple keypoints, and all have same color or does it mean that there are 3 instances and all have just 1 keypoint.
It's a bit ambiguous.

Where as with the current one it's bit straightforward,

`colors=["red", "blue", "green"]` for having 3 instances with same color for all keypoints
`colors=[["red"], ["blue"], ["green"]` for having 3 instances having color with just 1 keypoint.

Is there any edge case the current implementation does not handle well? I feel that either code are fine.

Reply:
FYI, I would love to use this utility on the Multi-pretrained weights example for Keypoint RCNN which is still pending. I don't want to express opinion over what's best here as I'm not too familiar with the KP model but I would love to see this unblocked. Francisco has a ton of experience with the specific model but if Oke has still concerns perhaps it's worth simplifying the utility (for example support 1 colour only) and open an issue to extend to all? Also I would like to see the PR containing an example on the Gallery but I see that's planned for follow up.

Reply:
@oke-aditya 
> Does this mean there are 3 instances with multiple keypoints, and all have same color or does it mean that there are 3 instances and all have just 1 keypoint.

Both are possible with broadcasting :-)

I think my proposal might be the "transposed" of yours. My specification is that we broadcast the colors to be of the same shape as the keypoints (so `[N, k]`). You can call `colors.T` in my proposal (after broadcasting) to make it compatible with yours.

Reply:
:sweat_smile: Do you suggest refactoring the current one? Or staying with the same :sweat_smile: 
I will leave the decision to you @fmassa :)  Seems like we have 2 valid and equivalent solutions.

Reply:
@oke-aditya Because it's been over 3 months since this PR has started (and that's not your problem, it's on us; thankfully you are a regular contributor and have understanding!), I think we should wrap it up and merge the potion for which there is consensus. As long as we ensure there will be no BC-breaking modification, we can always revisit it on a follow up PR.

In that direction, I recommend 2 ways to move forward:
1. If you have no major concerns about the API, the user-friendliness etc. Adopt @fmassa's proposal. 
2. If you are not convinced about the proposal, just support a single colour (1 string param) for all keypoints. 

In both cases, I would prefer if you added an example in the gallery folder. You can do it either here or on a follow up shortly after.

Let's wrap this up as this is a great addition and the community is missing out.

Reply:
I can go ahead with @fmassa's proposal :) I'm convinced by his idea.  Let me give a run with his idea and should we face not any bugs / problems. We can merge it

I will write up a gallery example in follow up PR very soon. I do have the code ready (which needs slight change for the adopting)
 
I guess there are a few aspects that need to be worked upon which won't be BC Breaking. 
Supporting Line color, not drawing with default white if None is passed (Similar to #4528).
Also should we plot the lines above the circle or vice-versa. I will create a ticket where we can track down these.